#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/Курс Дата аналітик /Data/supermarket_sales.csv')

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.groupby('Branch')['gross income'].sum().sort_values(ascending = False)

Branch
C    5265.1765
A    5057.1605
B    5057.0320
Name: gross income, dtype: float64

Найприбутковішим є філіал С.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
branch_income = df.groupby(['Branch','City'])['gross income'].sum()
branch_income

Branch  City     
A       Yangon       5057.1605
B       Mandalay     5057.0320
C       Naypyitaw    5265.1765
Name: gross income, dtype: float64

In [5]:
max_profit = branch_income.max()
max_profit

5265.1765

В місті Naypyitaw	знаходиться найприбутковіший філіал. Проте різниця між містами незначна, тому не впевнена, що варто відкривати там ще один магазин. Можливо це навпаки гірше вплине на прибуток в цьому місті (затрати будуть на +1 магазин).
Думаю треба дослідити наскільки часто це місто та цей філіал лідирують.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [36]:
invoices_by_category = df.pivot_table(index = 'Branch', columns = 'Product line', values = 'Invoice ID', aggfunc = 'count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [40]:
invoices_by_category['Electronic accessories'].max() #найбільше інвойсів

60

In [42]:
invoices_by_category['Electronic accessories'].idxmax() #в якому філіалі

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**Питання:**

1. Який максимальний,мінімальний та середній рейтинг ставлять покупці в місті `"Naypyitaw"`?

2. Де найнижча `"cogs"`?

3. Якого `"Customer type"` більше в місті `"Naypyitaw"`?

**1 питання**

Який максимальний, мінімальний та середній рейтинг ставлять покупці в місті `"Naypyitaw"`?

In [56]:
df['Rating'][df['City'] == 'Naypyitaw'].max() #максимальний рейтинг

10.0

In [57]:
df['Rating'][df['City'] == 'Naypyitaw'].min() #мінімальний рейтинг

4.0

In [65]:
df['Rating'][df['City'] == 'Naypyitaw'].mean() #середній рейтинг

7.072865853658537

In [66]:
naypyitaw_ratings = df['Rating'][df['City'] == 'Naypyitaw']
with pd.option_context("display.precision", 2):
  display(naypyitaw_ratings.describe())

  #тут коли буду вміти візуазізувати ще додам графік, і буде на графіку красивіше та зрозуміліше персантиль

count    328.00
mean       7.07
std        1.70
min        4.00
25%        5.60
50%        7.10
75%        8.50
max       10.00
Name: Rating, dtype: float64

**Висновок до першого питання:**

Ми бачимо, що рейтинг в місті `Naypyitaw` середнім значенням має - 7.07 бала. Це достатньо високий рейтинг.

Також з `персантилю` (та майбутнього графіку) ми бачимо, що половина всіх відгуків на цьому значенні та вище. А значить тільки половина клієнтів задоволені магазинами(ставлять 7+балів).

Нам потрібно провести додаткове опитування критеріїв, які подобаються(не подобаються) клієнтам.
Потім на основі цих даних запровадити зміни, щоб відсоток середньої та вище оцінки сягав більшого значення.

**2 питання**

Де найнижча `"cogs"`?

In [67]:
df.groupby('cogs').min()

Invoice ID Branch       City Customer type  Gender  \
cogs                                                          
10.17   784-21-9238      C  Naypyitaw        Member    Male   
12.09   308-39-1707      A     Yangon        Normal  Female   
12.54   279-62-1445      C  Naypyitaw        Member  Female   
12.78   192-98-7397      C  Naypyitaw        Normal    Male   
13.98   236-86-3015      C  Naypyitaw        Member    Male   
...             ...    ...        ...           ...     ...   
973.80  303-96-2227      B   Mandalay        Normal  Female   
975.00  751-41-9720      C  Naypyitaw        Normal    Male   
985.20  283-26-5248      C  Naypyitaw        Member  Female   
989.80  687-47-8271      A     Yangon        Normal    Male   
993.00  860-79-0874      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
cogs                                                                    
10.17     Sports and travel       10.17         1   0.5085    10.6785   
12.09   Fashion accessories       12.09         1   0.6045    12.6945   
12.54   Fashion accessories       12.54         1   0.6270    13.1670   
12.78   Fashion accessories       12.78         1   0.6390    13.4190   
13.98    Home and lifestyle       13.98         1   0.6990    14.6790   
...                     ...         ...       ...      ...        ...   
973.80   Home and lifestyle       97.38        10  48.6900  1022.4900   
975.00   Home and lifestyle       97.50        10  48.7500  1023.7500   
985.20   Food and beverages       98.52        10  49.2600  1034.4600   
989.80  Fashion accessories       98.98        10  49.4900  1039.2900   
993.00  Fashion accessories       99.30        10  49.6500  1042.6500   

             Date   Time      Payment  gross margin percentage  gross income  \
cogs                                                                           
10.17    2/7/2019  14:15         Cash                 4.761905        0.5085   
12.09   1/26/2019  18:19  Credit card                 4.761905        0.6045   
12.54   2/21/2019  12:38         Cash                 4.761905        0.6270   
12.78    1/8/2019  14:11      Ewallet                 4.761905        0.6390   
13.98    2/4/2019  13:38      Ewallet                 4.761905        0.6990   
...           ...    ...          ...                      ...           ...   
973.80   3/2/2019  17:16      Ewallet                 4.761905       48.6900   
975.00  1/12/2019  16:18      Ewallet                 4.761905       48.7500   
985.20  1/30/2019  20:23      Ewallet                 4.761905       49.2600   
989.80   2/8/2019  16:20  Credit card                 4.761905       49.4900   
993.00  2/15/2019  14:53  Credit card                 4.761905       49.6500   

        Rating  
cogs            
10.17      5.9  
12.09      8.2  
12.54      8.2  
12.78      9.5  
13.98      9.8  
...        ...  
973.80     4.4  
975.00     8.0  
985.20     4.5  
989.80     8.7  
993.00     6.6  

[990 rows x 16 columns]

**Висновок до другого питання:**

Ми бачимо, що найменший `cogs` обійшовся в 10.17. При цьому якщо порівняти `gross margin percentage` то він всюди однаковий.

Якщо говорити про відкриття нових магазині ще, то думаю варто відкривати аналоги з найменшим `cogs`, щоб менше грошей заморожувати в бізнесі. При цьому отримувати такий же `gross margin percentage`.



**3 питання**

Якого `"Customer type"` більше в місті `'Naypyitaw'`?

In [81]:
df[df['City'] == 'Naypyitaw']['Customer type'].value_counts().idxmax()

'Member'

In [84]:
df[df['City'] == 'Naypyitaw']['Customer type'].value_counts(normalize = True)

Customer type
Member    0.515244
Normal    0.484756
Name: proportion, dtype: float64

**Висновок на третє питання:**

Ми бачимо, що у місті `Naypyitaw` користувачів `Member` більше, ніж `Normal `. Проте ця різниця в кількості невелика.

Можна додатково дослідити чи роблять клієнти `Member `більші покупки за чеком, чи ні.

Якщо так, то тоді варто переводити `Normal `клієнтів через різні маркетингові інструменти в інший тип.
